pkg/collector/corechecks/systemd/systemd.go

```diff
@@ -0,0 +1,448 @@
+	return c.SystemState()
+}
+
+func (s *defaultSystemdStats) ListUnits(conn *dbus.Conn) ([]dbus.UnitStatus, error) {
+	return conn.ListUnits()
+}
+
+func (s *defaultSystemdStats) GetUnitTypeProperties(c *dbus.Conn, unitName string, unitType string) (map[string]interface{}, error) {
+	return c.GetUnitTypeProperties(unitName, unitType)
+}
+
+func (s *defaultSystemdStats) TimeNanoNow() int64 {
+	return time.Now().UnixNano()
+}
+
+// Run executes the check
+func (c *SystemdCheck) Run() error {
+	sender, err := aggregator.GetSender(c.ID())
+	if err != nil {
+		return err
+	}
+
+	conn, err := c.getDbusConn(sender)
+	if err != nil {
+		return err
+	}
+	defer c.stats.CloseConn(conn)
+
+	err = c.submitMetrics(sender, conn)
+	if err != nil {
+		return err
+	}
+	sender.Commit()
+	return nil
+}
+
+func (c *SystemdCheck) getDbusConn(sender aggregator.Sender) (*dbus.Conn, error) {
+	conn, err := c.stats.NewConn()
+	if err != nil {
+		newErr := fmt.Errorf("Cannot create a connection: %v", err)
+		sender.ServiceCheck(canConnectServiceCheck, metrics.ServiceCheckCritical, "", nil, newErr.Error())
+		return nil, newErr
+	}
+
+	systemStateProp, err := c.stats.SystemState(conn)
+	if err != nil {
+		newErr := fmt.Errorf("Err calling SystemState: %v", err)
+		sender.ServiceCheck(canConnectServiceCheck, metrics.ServiceCheckCritical, "", nil, newErr.Error())
+		return nil, newErr
+	}
+	sender.ServiceCheck(canConnectServiceCheck, metrics.ServiceCheckOK, "", nil, "")
+
+	serviceCheckStatus := metrics.ServiceCheckUnknown
+	systemState, ok := systemStateProp.Value.Value().(string)
+	if ok {
+		status, ok := systemdStatusMapping[systemState]
+		if ok {
+			serviceCheckStatus = status
+		}
+	}
+	sender.ServiceCheck(systemStateServiceCheck, serviceCheckStatus, "", nil, fmt.Sprintf("Systemd status is %v", systemStateProp.Value))
+	return conn, nil
+}
+
+func (c *SystemdCheck) submitMetrics(sender aggregator.Sender, conn *dbus.Conn) error {
+	units, err := c.stats.ListUnits(conn)
+	if err != nil {
+		return fmt.Errorf("Error getting list of units: %v", err)
+	}
+
+	c.submitCountMetrics(sender, units)
+
+	loadedCount := 0
+	for _, unit := range units {
+		if unit.LoadState == unitLoadedState {
+			loadedCount++
+		}
+		if !c.isMonitored(unit.Name) {
+			continue
+		}
+		tags := []string{"unit:" + unit.Name}
+		sender.ServiceCheck(unitStateServiceCheck, getServiceCheckStatus(unit.ActiveState), "", tags, "")
+
+		c.submitBasicUnitMetrics(sender, conn, unit, tags)
+		c.submitPropertyMetricsAsGauge(sender, conn, unit, tags)
+	}
+
+	sender.Gauge("systemd.unit.loaded.count", float64(loadedCount), "", nil)
+	return nil
+}
+
+func (c *SystemdCheck) submitBasicUnitMetrics(sender aggregator.Sender, conn *dbus.Conn, unit dbus.UnitStatus, tags []string) {
+	unitProperties, err := c.stats.GetUnitTypeProperties(conn, unit.Name, dbusTypeMap[typeUnit])
+	if err != nil {
+		log.Errorf("Error getting unit unitProperties: %s", unit.Name)
```

Comment:
Applies to other `log` calls as well during a run of the check: if you want this message to show up as a warning on the status page + on the host in the DD infra list + get logged with a `Warn` level, you can use `c.Warnf` which is implemented in `CheckBase` (see https://github.com/DataDog/datadog-agent/blob/master/pkg/collector/corechecks/checkbase.go#L105-L119). You'll want to use `c.Warnf` for the warnings that the user should definitely see on DD and in the status page.

In general, I'd also recommend using lower log levels than `Error` (rather Warn or Info), except for very serious errors.

Reply:
Yes, Error is a bit too high for many cases in this integration.

Fixed. I lowered the log level of all non serious errors.